tests/test_compile_kwargs.py

```diff
@@ -0,0 +1,557 @@
+    string is passed as an optimizer.
+    """
+
+    X, y = make_classification()
+
+    optimizer = "sgf"  # sgf is not a loss
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        loss="binary_crossentropy",
+    )
+    with pytest.raises(ValueError, match=f"Unknown optimizer: {optimizer}"):
+        est.fit(X, y)
+
+
+def test_compiling_of_routed_parameters():
+    """Tests that routed parameters
+    can themselves be compiled.
+    """
+
+    X, y = make_classification()
+
+    class Foo:
+        got = dict()
+
+        def __init__(self, **kwargs):
+            self.got = kwargs
+
+    class MyLoss:
+        def __init__(self, param1):
+            self.param1 = param1
+            self.__name__ = str(id(self))
+
+        def __call__(self, y_true, y_pred):
+            return losses_module.binary_crossentropy(y_true, y_pred)
+
+    est = KerasClassifier(
+        model=get_model,
+        loss=MyLoss,
+        loss__param1=[Foo, Foo],
+        loss__param1__foo_kwarg=1,
+        loss__param1__0__foo_kwarg=2,
+    )
+    est.fit(X, y)
+    assert est.model_.loss.param1[0].got["foo_kwarg"] == 2
+    assert est.model_.loss.param1[1].got["foo_kwarg"] == 1
+
+
+@pytest.mark.parametrize(
+    "loss", (losses_module.BinaryCrossentropy, "BinaryCrossentropy",),
+)
+@pytest.mark.parametrize("n_outputs_", (1, 2))
+def test_loss(loss, n_outputs_):
+    """Tests compiling of single loss
+    using routed parameters.
+    """
+
+    X, y = make_classification()
+    y = np.column_stack([y for _ in range(n_outputs_)]).squeeze()
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer="sgd",
+        loss=loss,
+        loss__name="custom_name",
+    )
+    est.fit(X, y)
+    assert str(loss) in str(est.model_.loss) or isinstance(
+        est.model_.loss, loss
+    )
+
+
+def test_loss_invalid_string():
+    """Tests that a ValueError is raised when an unknown
+    string is passed as a loss.
+    """
+
+    X, y = make_classification()
+
+    loss = "binary_crossentropr"  # binary_crossentropr is not a loss
+
+    est = KerasClassifier(
+        model=get_model, num_hidden=20, optimizer="sgd", loss=loss,
+    )
+    with pytest.raises(ValueError, match=f"Unknown loss function: {loss}"):
+        est.fit(X, y)
+
+
+def test_loss_uncompilable():
+    """Tests that a TypeError is raised when a loss
+    that is not compilable is passed routed parameters.
+    """
+
+    X, y = make_classification()
+
+    loss = losses_module.binary_crossentropy
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer="sgd",
+        loss=loss,
+        loss__from_logits=True,
+    )
+    expected_param = {"from_logits": True}
+    with pytest.raises(
+        TypeError,
+        match=f'TypeError: "{str(loss)}" object of type "{type(loss)}" could not be called with'
+        f" parameters {expected_param}",
```

Comment:
I chose this wording as opposed to `is not callable` because I felt that it might be confusing if we tell the user that a function is not callable. @stsievert  please feel free to suggest alternative wording if you have anything that might be clearer.

Reply:
Why not add an extra branch in the if-statement of `compile_with_params` with `isinstance(items, function) or callable(f): return items`?

Reply:
With the goal being to raise a different error for callables that are not classes?

Reply:
You mentioned that you avoided "is not callable" in the error message because "it might be confusing if we tell the user that a function is not callable." This avoid that – now `compile_with_params` can't raise the error message for functions/callables.

Reply:
Nevermind, that's a poorly worded comment. Let me think on better wording/code.

Reply:
But I think we _do_ want it to raise an error if the user routes a parameter to function/callable. The issue is that we want to instantiate classes with routed parameters not call functions (I don't think there is any scenario where we call a function with routed parameters). So really the error should be something along the lines of "you passed routed parameters to something that is not an instantiable class", but I'm not sure that I like that specific wording.

Reply:
> Nevermind, that's a poorly worded comment. Let me think on better wording/code.

Thanks!

Reply:
Maybe this error message?

> TypeError: "{loss}" object of type "{type(loss)}" does not accept parameters because it's not a class. However, it received parameters of {passed_param_name}={expected_params}.

This would require some work to get `passed_param_name`.

Reply:
Where is `passed_param_name` and `expected_params` coming from? I was using `expected_params` in testing just to adjust for removal of the routing prefixes, but I feel like the error (raised within `compile_with_params`) would be:

> TypeError: "{str(item)}" object of type "{type(item)}" does not accept parameters because it's not a class. However, it received parameters "{kwargs}"

Reply:
That error message looks correct: I used `expected_params` as a placeholder because I thought this comment was in `compile_with_params`.

> This would require some work to get `passed_param_name`.

This work will likely involve passing the prefix down each level of recursion.

Reply:
Do you think the complexity that that would require is worth it? I feel like it would only be valuable when users are using the same arguments at different levels of nesting, otherwise it should be fairly obvious where the mistake happened.

Reply:
Let's leave it then. I'll resolve this conversation.